### ISSUE TITLE
PPoPP 26 | HelixPipe: Efficient Distributed Training of Long Sequence Transformers with Attention Parallel Pipeline Parallelism

### ISSUE THREAD

**Issue by Dszdsxc0405** [AI-Infra-Team/awesome-papers#162]
## Paper Infomation

**Title:** HelixPipe: Efficient Distributed Training of Long Sequence Transformers with Attention Parallel Pipeline Parallelism
**Authors:** Geng Zhang, Shenggan Cheng, Xuanlei Zhao, Ziming Liu, Yang You
**Institudes:**  National University of Singapore Singapore, Singapore
**Paper:** [[2507.00394] HelixPipe: Efficient Distributed Training of Long Sequence Transformers with Attention Parallel Pipeline Parallelism](https://arxiv.org/abs/2507.00394)
**Github Note Link:** [zxgx/Megatron-LM: Ongoing research training transformer models at scale](https://github.com/zxgx/Megatron-LM/tree/dev)
**Author Page:** None

## Background

In recent years, Transformer models have achieved groundbreaking progress in code generation, image creation, and scientific computing. Their success largely stems from the continuous enhancement of the model's ability to understand and generate long-sequence contexts. However, when sequence length expands from the common thousands to tens or even hundreds of thousands, model training faces severe systemic challenges. The core issue lies in the quadratic growth of the attention mechanism's computational complexity with sequence length, coupled with the explosive increase in memory demand for storing intermediate activations, which far exceeds the capabilities of a single computing device.

To address this, industry and academia commonly employ a combination of parallelization strategies for distributed training. Among these, **pipeline parallelism**—a scheme that splits the model layer-wise across different devices for pipelined execution—is crucial for reducing single-device memory pressure and extending model depth. 

Classic pipeline parallel methods, such as **1F1B** widely used in frameworks like **Megatron-LM** and its improved variant **Zero Bubble**, have proven effective for shorter sequence lengths (e.g., within 2048). They partition the Transformer model into several consecutive layer blocks, each assigned as a "stage" to a different GPU. During training, data batches are further divided into micro-batches and injected into the pipeline sequentially. 

## Problems and Motivation

Although existing methods perform well on short sequences, their performance deteriorates rapidly when facing long-sequence training, primarily due to two interrelated fundamental issues.

**First, the pipeline bubble is dominated by the drastically inflated attention computation, leading to low computational efficiency.**  
Because traditional methods partition at the granularity of an **entire Transformer layer**, the bubble duration is proportional to the execution time of one layer. Due to its quadratic complexity, the attention mechanism quickly becomes the absolute dominant part of per-layer computation as sequence length increases.  
**Figure 3** in the paper clearly illustrates this trend:

> When the sequence length reaches 128K, attention computation occupies over 80% of the forward execution time for a single layer.
<div align="center">
    <img width="448" height="335" alt="Image" src="https://github.com/user-attachments/assets/402e1266-e1b9-4e15-ac94-b6cc384007aa" />
</div>

Consequently, in the bubble formulas for 1F1B or Zero Bubble, the term representing attention time, `t_attn`, becomes the decisive factor. This causes the bubble time to increase significantly with sequence length, wasting a large number of GPU cycles on waiting.

<div align="center">
    <img width="365" height="45" alt="Image" src="https://github.com/user-attachments/assets/16e856a0-4614-4abb-99e5-9808ca205a1c" />
    <img width="365" height="45" alt="Image" src="https://github.com/user-attachments/assets/d5f860cd-e39c-4874-870e-28ee1fde7263" />
</div>

---

**Second, the extreme imbalance in activation memory across pipeline stages limits the trainable sequence length.**  

Under the 1F1B scheduling policy, the earliest pipeline stages need to preserve activations for the greatest number of micro-batches still "in-flight" within the pipeline, in preparation for backward propagation. **Formula 2** in the paper reveals that the activation memory required by the first stage is always as high as `16bshL`, independent of the total number of pipeline stages.  

<div align="center">
    <img width="365" height="45" alt="Image" src="https://github.com/user-attachments/assets/71f192fb-22ce-43d0-aeb7-c09ea375bba4" />
</div>

As shown in **Figure 4**, when training a 13B-parameter model with a sequence length of 128K using 8 pipeline stages, the activation memory demand of the first and second stages far exceeds the 80GB GPU memory capacity, while later stages have substantial idle memory. This "uneven distribution" of memory causes the overall training scale to be constrained by the most strained stage.  

<div align="center">
    <img width="421" height="282" alt="Image" src="https://github.com/user-attachments/assets/599f0734-fb60-49b7-b44d-a9dd5d1711b6" />
</div>

Although the Zero Bubble method theoretically optimizes the computation flow, it does not change the essential requirement to fully save activations for all micro-batches until the final parameter gradient calculation. Therefore, its peak memory remains the same as that of 1F1B, unable to alleviate this memory bottleneck.

<div align="center">
    <img width="365" height="45" alt="Image" src="https://github.com/user-attachments/assets/fddc74c1-de0a-4403-8afc-130621d659e8" />
</div>

## Observation and Opportunity

Through in-depth analysis of the above problems, the authors of the paper identified a key characteristic, thereby finding an opportunity to break through the困境: **In a Transformer layer, the computationally most intensive attention module itself does not contain any trainable parameters.** This means that the results of attention computation (activations) only need to be temporarily retained for subsequent layer computations and gradient backpropagation, but the computation process itself does not depend on local model weights.

<div align="center">
    <img width="439" height="243" alt="Image" src="https://github.com/user-attachments/assets/5126f103-c0ce-4358-bcff-eeb38a84fa88" />
</div>

This observation brings fundamental inspiration: Since attention computation is so expensive yet parameter-independent, can it be **decoupled** from the traditional, layer-bound pipeline execution mode? Furthermore, can attention computations for different micro-batches be **executed in parallel** on different pipeline stages, thereby removing them from the serial, blocking pipeline bubble? At the same time, this decoupling also creates new possibilities for redesigning micro-batch scheduling to balance the memory load across stages.

### 1. Attention Parallel Partitioning

**Function: Transform serial bottleneck into parallel opportunity, accelerating attention computation**

Attention Parallel Partitioning: Transforming a Serial Bottleneck into a Parallel Opportunity. Traditional pipeline parallelism adopts a "whole-layer contracting" approach, treating a complete Transformer layer (including pre-attention, attention, and post-attention computation phases) as an indivisible unit of work assigned to one pipeline stage.  

**As shown in the left half of Figure 5**, in a simple scenario with 2 stages and 2 micro-batches, when Stage 0 processes the entire layer for micro-batch 1, Stage 1 remains completely idle. Since attention computation takes the longest time (the paper assumes it occupies 50% of single-layer computation time), this "layer-granularity" partitioning causes the most expensive computational resources to be executed sequentially, forming a large pipeline bubble.

HelixPipe breaks this rigid pattern. Seizing the key characteristic that "the attention module has no model parameters," it splits a Transformer layer into three independent processing segments: **Pre-Attention, Attention, and Post-Attention**. Its innovation lies in introducing a **"Helical" allocation rule**: it distributes the parameterized pre-attention and post-attention parts across stages in an interleaved manner according to layers, while dynamically scheduling attention computations of different micro-batches to different stages.

<div align="center">
    <img width="428" height="246" alt="Image" src="https://github.com/user-attachments/assets/785f09b3-ac91-4cdb-aa67-ae11646831ab" />
</div>

**As illustrated in the right half of Figure 5**, HelixPipe's allocation rule determines that: the attention computation for micro-batch 1 is assigned to Stage 1, while the attention for micro-batch 2 remains on Stage 0. When Stage 1 is performing the attention computation for MB1 lasting 3 time units, Stage 0, after completing the pre-attention for MB2, also begins a 3-unit attention computation for MB2. Thus, during the critical 3-unit period, the two GPUs achieve complete parallel processing of the most time-consuming task.  

### 2. Double-Folded First-In-Last-Out (Filo) Scheduling

**Function: Overlap computation with communication, balance memory load**

However, such fine-grained partitioning inevitably leads to a significant increase in the frequency of data communication between stages (e.g., pre-attention results need to be sent to the stage responsible for attention computation), potentially creating a new communication bottleneck. To address this, HelixPipe designs a **Double-Folded First-In-Last-Out (Filo) Scheduling** strategy, aiming to solve both the communication latency and memory imbalance problems simultaneously.

**First,** **the essence of its Filo nature** requires all micro-batches to sequentially complete the forward propagation of all layers first, and then execute backward propagation in the exact reverse order. **As shown in Figure 2b**, this strategy ensures that before backward propagation begins, the number of activation values each pipeline stage needs to save for all "in-flight" micro-batches is the same (equal to the number of pipeline stages, p). This fundamentally solves the imbalance problem in traditional 1F1B where early stages suffer from excessive memory pressure.
<div align="center">
    <img width="800" height="264" alt="Image" src="https://github.com/user-attachments/assets/1601f80e-73cf-4058-bffc-1b108dd10e2f" />
</div>

**Second, to cope with the intensive communication introduced by partitioning, the scheduling is upgraded to a double-folded mode**. Its core idea is to collaboratively schedule two micro-batches as one "fold" at a time. **As shown in Figure 6b**, key operation overlap is achieved: while Stage 1 is computing attention for one micro-batch, Stage 0 can simultaneously compute attention for another micro-batch. The data communication between them (e.g., sending back the attention results of the previous micro-batch) is cleverly hidden within this parallel computation time.
<div align="center">
    <img width="444" height="386" alt="Image" src="https://github.com/user-attachments/assets/bf8f15f7-91f5-4019-8b69-d14e7b2511a9" />
</div>

The comparison in **Figure 7** clearly shows that although the double-folded strategy slightly increases the initial bubble because it processes two micro-batches at once, it successfully covers all subsequent communication operations (the purple blocks in Figure 7b) completely under the computation operations (green blocks), thereby eliminating the waiting delay caused by communication overall.
<div align="center">
    <img width="800" height="271" alt="Image" src="https://github.com/user-attachments/assets/4b7fbb86-68cd-43a2-abc8-b6c1f135fb31" />
</div>

### 3. Attention-Excluded Recomputation and Chunked MLP

**Function: Break through the memory bottleneck, enable longer training sequences**

After addressing computational efficiency and memory balance, HelixPipe further supports training with even longer sequences through memory optimization. Its strategy is based on a key observation: under extremely long sequences, the vast majority of computation time in a layer (e.g., over 85% at 128K) is occupied by attention, while the computational proportion of the pre-attention and post-attention operations is very small (Figure 3).

Therefore, HelixPipe adopts a **counter-intuitive attention-excluded recomputation strategy**. Unlike traditional selective recomputation which saves memory for attention, HelixPipe **does not store all intermediate activations for the pre-attention and post-attention parts**. It only stores the input/output of the attention module and the necessary input for the post-attention part. Before backward propagation, it uses this saved data to temporarily recompute the intermediate activations required for the pre-attention and post-attention parts. 

According to the **quantitative analysis in Table 2**, this strategy, at the cost of a small amount of computation time (the coefficient in the bubble time formula increases from 6 to 8), drastically reduces the activation memory required per layer from `16bsh` to `4bsh`, achieving a **4x memory saving**.
<div align="center">
    <img width="463" height="213" alt="Image" src="https://github.com/user-attachments/assets/0c29b247-6a80-4fbc-80ea-5c26cf581c59" />
</div>

Furthermore, the introduction of recomputation, double-folded scheduling, coupled with the huge tensors generated by long sequences themselves, can easily lead to severe memory fragmentation during MLP layer computation. To address this, HelixPipe employs a **Chunked MLP** technique. Instead of processing the entire massive hidden state tensor at once, it splits it along the sequence dimension into fixed-size chunks, performing computation and aggregation chunk by chunk. This approach transforms a single irregular large memory allocation and computation into multiple regular small-chunk operations, fundamentally alleviating GPU memory fragmentation issues.

## Experiments and Results

| Item | Specific Configuration |
| :--- | :--- |
| **Test Platform** | **H20 Cluster**: 8×H20 GPUs per node, 4×IB NDR HCAs (200 Gbps/port) <br> **A800 Cluster**: 8×A800 GPUs per node, 4×IB HDR HCAs (100 Gbps/port) |
| **Workload** | **Model**: GPT-3 architecture, parameter scales 1.3B, 3B, 7B (configuration see Table 3) <br> **Sequence Length**: 32K, 64K, 96K, 128K <br> **Data**: Synthetic dataset, full sequence length, no padding <br> **Parallel Config**: 1 pipeline stage = 1 node; intra-node uses Megatron sequence parallelism (degree 8) |
| **Baseline Methods** | **1F1B**: The most widely adopted pipeline parallel baseline <br> **ZB1P**: An efficient zero-bubble pipeline variant <br> **AdaPipe**: An adaptive recomputation and partitioning method for memory optimization |
| **Evaluation Metrics** | **Throughput**: Normalized training throughput per iteration (average of 10 iterations after warm-up) <br> **Memory**: Peak memory usage recorded by the PyTorch caching allocator |

### 1. Computational Efficiency and Scalability
As shown in Figure 8, HelixPipe achieves the highest throughput in the vast majority of configurations:
- **On the H20 cluster, using 8 nodes** (i.e., 8 pipeline stages) to train the 7B model with a 128K sequence length, HelixPipe achieves a **26% throughput improvement** compared to the best baseline.
- **On the A800 cluster** under the corresponding configuration, the improvement is **13%**. Analysis shows that HelixPipe's advantage becomes more pronounced as sequence length, model scale, and the number of pipeline stages increase, precisely validating that its design effectively tackles the attention-dominated computation and memory bottlenecks brought by long sequences.

It is noteworthy that on the A800 cluster, the performance improvement for shorter sequences (e.g., 32K) is limited. This stems from its higher computational power making attention computation faster, while its relatively lower communication bandwidth makes it difficult for the communication in the double-folded schedule to be fully overlapped.
<div align="center">
   <img width="800" height="481" alt="Image" src="https://github.com/user-attachments/assets/0fd8a439-3aa4-48a3-a53f-8c8251c19353" />
</div>

### 2. Impact of Communication

The effectiveness of the double-folded Filo schedule depends on whether the attention computation can overlap the underlying communication.

**Specifically, on the A800 cluster**, the attention computation speed for a 32K sequence length is faster than the inter-node communication. Therefore, the communication introduced by the double-folded Filo schedule cannot be fully overlapped by the attention computation, causing subsequent computation delays.
Nonetheless, for longer sequence lengths, HelixPipe can still achieve decent speedups on the A800 cluster because the communication will be completely overlapped by the computation.

On the other hand, **communication on the H20 cluster** is fast enough to be perfectly overlapped by the attention computation, which means HelixPipe can scale to any number of nodes on this cluster.
<div align="center">
   <img width="362" height="303" alt="Image" src="https://github.com/user-attachments/assets/7fb2b1f8-60ca-479f-8f75-73c97bd266ee" />
</div>

### 3. Memory Footprint
**As shown in Figure 10**, HelixPipe exhibits the lowest and most balanced peak memory usage across stages. Taking training a 3B model with a 128K sequence length across 8 stages as an example: 1F1B's memory usage is severely skewed, with enormous memory pressure on early stages; ZB1P has consistent peak memory across stages but the value is very high; AdaPipe attempts to balance but is limited by long sequences and not fully effective. Only HelixPipe significantly reduces the absolute memory value while achieving balanced load across all stages.
<div align="center">
   <img width="362" height="313" alt="Image" src="https://github.com/user-attachments/assets/c070358f-2920-4436-aebd-854d1998487f" />
</div>

### 4. Ablation Study

The impact of the **Attention-Excluded Recomputation strategy** is verified (Figure 11). This strategy incurs up to a 20% throughput loss on shorter sequences (32K, 64K) because the pre-attention and post-attention computations still occupy a considerable proportion at those lengths. However, as the sequence grows to 96K and 128K, the proportion of additional computational overhead it introduces decreases sharply, while the **4x reduction in activation memory** it brings becomes the key to supporting training with ultra-long sequences.
<div align="center">
   <img width="413" height="272" alt="Image" src="https://github.com/user-attachments/assets/e3832a46-eb45-4b1c-b64c-06d53f317d51" />
</div>
